LongWarOfTheChosen/Src/LW_Overhaul/Classes/UIScreenListener_ArmoryWeaponUpgrade_LW.uc

```diff
@@ -1,156 +1,436 @@
+						UpdateState.AddStateObject(XComHQ);
+
+						// If this is the only instance of this weapon in the inventory we'll just get back a non-updated state.
+						// That's ok, StripWeaponUpgradesFromItem will create/add it if it's not already in the update state. If it
+						// is, we'll use that one directly to do the stripping.
+						XComHQ.GetItemFromInventory(UpdateState, ItemState.GetReference(), ItemState);
+						StripWeaponUpgradesFromItem(ItemState, XComHQ, UpdateState);
+						ItemState = XComGameState_Item(UpdateState.GetGameStateForObjectID(ItemState.ObjectID));
+						XComHQ.PutItemInInventory(UpdateState, ItemState);
+						`GAMERULES.SubmitGameState(UpdateState);
+					}
+				}
+			}
+		}
+
+		// Strip upgrades from weapons on soldiers that aren't active. These can all be batched in one state because
+		// soldiers maintain their equipped weapon, so there is no stacking of weapons to consider.
+		UpdateState = class'XComGameStateContext_ChangeContainer'.static.CreateChangeState("Strip Unequipped Upgrades");
+		XComHQ = XComGameState_HeadquartersXCom(UpdateState.CreateStateObject(class'XComGameState_HeadquartersXCom', XComHQ.ObjectID));
+		UpdateState.AddStateObject(XComHQ);
+		Soldiers = GetSoldiersToStrip(XComHQ, UpdateState);
+		for (idx = 0; idx < Soldiers.Length; idx++)
+		{
+			class'UIUtilities_Strategy'.static.GetWeaponUpgradeAvailability(Soldiers[idx], WeaponUpgradeAvailabilityData);
+			if (!WeaponUpgradeAvailabilityData.bCanWeaponBeUpgraded)
+			{
+				continue;
+			}
+
+			ItemState = Soldiers[idx].GetItemInSlot(eInvSlot_PrimaryWeapon, UpdateState);
+			if (ItemState != none && ItemState.HasBeenModified())
+			{
+				StripWeaponUpgradesFromItem(ItemState, XComHQ, UpdateState);
+			}
+		}
+
+		`GAMERULES.SubmitGameState(UpdateState);
+	}
+	if (LoadoutScreen != none)
+	{
+		LoadoutScreen.UpdateNavHelp();
 	}
 }
 
-simulated function UpdateCustomizationStripWeapon(UIPanel Panel)
+function StripWeaponUpgradesFromItem(XComGameState_Item ItemState, XComGameState_HeadquartersXCom XComHQ, XComGameState UpdateState)
 {
-	local UIArmory_WeaponUpgrade UpgradeScreen;
+	local int k;
+	local array<X2WeaponUpgradeTemplate> UpgradeTemplates;
+	local XComGameState_Item UpdateItemState, UpgradeItemState;
 
-	UpgradeScreen = ValidateScreen(Panel.Screen);
-	if (UpgradeScreen != none)
-		RefreshScreen(UpgradeScreen);
+	UpdateItemState = XComGameState_Item(UpdateState.GetGameStateForObjectID(ItemState.ObjectID));
+	if (UpdateItemState == none)
+	{
+		UpdateItemState = XComGameState_Item(UpdateState.CreateStateObject(class'XComGameState_Item', ItemState.ObjectID));
+		UpdateState.AddStateObject(UpdateItemState);
+	}
+
+	UpgradeTemplates = ItemState.GetMyWeaponUpgradeTemplates();
+	for (k = 0; k < UpgradeTemplates.length; k++)
+	{
+		UpgradeItemState = UpgradeTemplates[k].CreateInstanceFromTemplate(UpdateState);
+		UpdateState.AddStateObject(UpgradeItemState);
```

Comment:
This should no longer be necessary as I'm pretty sure `CreateInstanceFromTemplate()` uses the newer WOTC API, but I'd need to get on my XCOM 2 PC to check.
```suggestion
```